applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp

```diff
@@ -71,14 +75,49 @@ namespace Kratos
     //************************************************************************************
     //************************************************************************************
 
-    void BaseSolidElement::FinalizeSolutionStep( ProcessInfo& CurrentProcessInfo )
+    void BaseSolidElement::FinalizeSolutionStep( ProcessInfo& rCurrentProcessInfo )
     {
-        //         std::cout << "in TL: calling FinalizeSolutionStep" << std::endl;
-        for ( unsigned int i = 0; i < mConstitutiveLawVector.size(); i++ )
-            mConstitutiveLawVector[i]->FinalizeSolutionStep( GetProperties(),
-                    GetGeometry(),
-                    row( GetGeometry().ShapeFunctionsValues(  ), i ),
-                    CurrentProcessInfo );
+        // Create and initialize element variables:
+        const unsigned int number_of_nodes = GetGeometry().size();
+        const unsigned int dimension = GetGeometry().WorkingSpaceDimension();
+        const unsigned int strain_size = GetProperties().GetValue( CONSTITUTIVE_LAW )->GetStrainSize();
+            
+        KinematicVariables this_kinematic_variables(strain_size, dimension, number_of_nodes);
+        ConstitutiveVariables this_constitutive_variables(strain_size);
+
+        // Create constitutive law parameters:
+        ConstitutiveLaw::Parameters Values(GetGeometry(),GetProperties(),rCurrentProcessInfo);
+
+        // Set constitutive law flags:
+        Flags& ConstitutiveLawOptions=Values.GetOptions();
+        ConstitutiveLawOptions.Set(ConstitutiveLaw::COMPUTE_STRAIN, false);
+        ConstitutiveLawOptions.Set(ConstitutiveLaw::COMPUTE_STRESS, true);
+        ConstitutiveLawOptions.Set(ConstitutiveLaw::COMPUTE_CONSTITUTIVE_TENSOR, false);
+
+        Values.SetStrainVector(this_constitutive_variables.StrainVector);
+        
+        // Reading integration points
+        const GeometryType::IntegrationPointsArrayType& integration_points = GetGeometry().IntegrationPoints(  );
+        
+        // Displacements vector
+        Vector displacements;
+        GetValuesVector(displacements);
+            
+        // Reading integration points
+        for ( unsigned int point_number = 0; point_number < mConstitutiveLawVector.size(); point_number++ )
+        {
+            // Compute element kinematics B, F, DN_DX ...
+            CalculateKinematicVariables(this_kinematic_variables, point_number, integration_points);
+            
+            // Call the constitutive law to update material variables
+            mConstitutiveLawVector[point_number]->FinalizeMaterialResponse(Values, GetStressMeasure());
```

Comment:
btw, if anyone at @KratosMultiphysics/structural-mechanics understands this (@marandra maybe?) please comment